### PR TITLE
fix(renderer): fix jumping to the top when opening files.

### LINKED ISSUE #996: window is scrolled up after collapsing node
I was trying neo-tree on https://github.com/qmk/qmk_firmware for checking performance measurement.

When a lots of folders are expanded in `keyboards` folder, and I collapse one of them, the window is scrolled upwards and so the collapsed node jumps at the bottom.

Kapture 2023-06-18 at 01 02 37

I debugged a bit, and these are  the code responsible for the unnecessary jump in this case:

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/95776b694e752b0a8feb45e4307d4a3913491eff/lua/neo-tree/ui/renderer.lua#L515-L526

### LINKED ISSUE #1025: BUG: Neo-tree window "jumps" when opening files
### Did you check docs and existing issues?

- [X] I have read all the docs.
- [X] I have searched the existing issues.
- [X] I have searched the existing discussions.

### Neovim Version (nvim -v)

NVIM v0.10.0-dev-628+g3ecd45ded-Homebrew

### Operating System / Version

MacOS 14.0

### Describe the Bug

When browsing a code base with lots of directories opened (to fill up the neo-tree window and be able to scroll it), when I open a file that I have to scroll down to get to, the neo-tree window will jump to the top (see the video below). 

~~I've tracked the issue to `document_symbols`. When I don't have it enabled, no jumping occurs. The config below reproduces this problem.~~

### Screenshots, Traceback

https://github.com/nvim-neo-tree/neo-tree.nvim/assets/2929/3fd2831d-b301-426e-874a-9490d7c0e80e

### Steps to Reproduce

1. Open project
2. Expand a bunch of directories to fill up the screen
3. Open files that are close to the bottom of the neo-tree window

### Expected Behavior

The neo-tree window should not jump around when opening files. I managed to fix it by not enabling `document_symbols`. (I'm using LazyVim, and LazyVim enabled document_symbols).

### Your Configuration

```Lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
	vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
	vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath })
end
vim.opt.runtimepath:prepend(lazypath)

-- install plugins
local plugins = {
	"folke/tokyonight.nvim",
}

local neotree_config = {
	"nvim-neo-tree/neo-tree.nvim",
	dependencies = { "MunifTanjim/nui.nvim", "nvim-tree/nvim-web-devicons", "nvim-lua/plenary.nvim" },
	cmd = { "Neotree" },
	keys = {
		{ "<Leader>e", "<Cmd>Neotree<CR>" }, -- change or remove this line if relevant.
	},
	opts = {
		sources = { "filesystem", "buffers", "git_status", "document_symbols" },
	},
}

table.insert(plugins, neotree_config)
require("lazy").setup(plugins, {
	root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
```

### PR DESCRIPTION
**Pull request by georgeguimaraes**
When opening files you had to scroll down to find, the neo-tree window jumped to the top (see #1025).

This is because Neovim changed the way `nvim_buf_set_lines` works in https://github.com/neovim/neovim/commit/008154954791001efcc46c28146e21403f3a698b (https://github.com/neovim/neovim/pull/24824) and later in https://github.com/neovim/neovim/pull/24889

This change in Neovim caused some problems in plugins:
- https://github.com/nvim-telescope/telescope.nvim/issues/2667
- https://github.com/folke/which-key.nvim/pull/516

To ensure the window stays fixed, we have to use `winsaveview` and `winrestview` (wrapped in `M.position`) inside `render_tree`.

Fixes #1025 
Fixes #996

### CODE REVIEW

**Comment by georgeguimaraes** [comment]
Please @miversen33 can you try it and see if I made a mess?

**Comment by georgeguimaraes** [comment]
I have a fix for #996 that builds on top of this PR. I'll have my commit cleaned up and I'll test a few edge cases and I'll update this PR

**Comment by georgeguimaraes** [comment]
Ok, now [6e6c42f](https://github.com/nvim-neo-tree/neo-tree.nvim/pull/1274/commits/6e6c42f6e4dec58dcb22ea735962404fd9f22ee3) fixes #996 as well. I've tested some cases with it, but I am not confident of such change, even though it uses well-established functions `winsaveview/winrestview`.

If you all feel more comfortable, I can remove this commit and we can just focus on #1025.

**Comment by miversen33** [comment]
I'll pull this down and review the branch and code this evening, I have a nice long drive ahead of me this afternoon

**Comment by miversen33** [comment]
I left some notes on this. The overall gist seems good though I will not be the one to sign off on this. I will leave that to either @pysan3  or @cseickel  as they are far more active in neo-tree core than myself. They will in theory be able to see potential side effects from this code that I might miss.

File: lua/neo-tree/ui/renderer.lua [hunk 1]
```diff
@@ -640,12 +640,12 @@ end
...
   save = function(state)
     if state.tree and M.window_exists(state) then
-      local success, node = pcall(state.tree.get_node, state.tree)
-      if success and node then
-        _, state.position.node_id = pcall(node.get_id, node)
-        local win_state = vim.fn.winsaveview()
-        state.position.topline = win_state.topline
-      end
+      local win_state = vim.api.nvim_win_call(state.winid, vim.fn.winsaveview)
+      log.trace("win_state: " .. vim.inspect(win_state))
```
Comment (pysan3) [inline lua/neo-tree/ui/renderer.lua:644]: Just a minor pick; I don't think it is good to leave `vim.inspect` inside the codebase, as the log.debug below already gives good info.

**Review by pysan3 — CHANGES_REQUESTED** [review]

**Comment by pysan3** [comment]
The change looks awesome on my end! Thanks a lot!

I'll pass it to @cseickel now.

**Review by cseickel — APPROVED** [review]
Looks reasonable to me and since @miversen33 and @pysan3 also reviewed I'll skip testing it myself.

Thanks!

**Comment by OliverChao** [comment]
Sorry for mentioning that this pr seems breaking the `open` action in `Neotree reveal`.

Open one file and run `Neotree reveal`, then run `<cr>` to `open` the directory  of that file.

After running `<CR>`,  it jumps to the file location not toggle the directory.

**Comment by cseickel** [comment]
> Sorry for mentioning that this pr seems breaking the `open` action in `Neotree reveal`.
> 
> Open one file and run `Neotree reveal`, then run `<cr>` to `open` the directory of that file.
> 
> After running `<CR>`, it jumps to the file location not toggle the directory.

That is true. Beyond that, any action you take will result in the previously revealed file being focused. That includes opening other files. Whatever the last revealed file is will be permanently stuck and refocused after every render.

I will make this a new issue.
